src/services/refactors/inlineFunction.ts

```diff
@@ -0,0 +1,389 @@
+        if (!info) return undefined;
+        const { declaration, usages, selectedUsage } = info;
+
+        switch (actionName) {
+            case inlineAllActionName:
+                return { edits: getInlineAllEdits(context, declaration, usages) };
+            case inlineHereActionName:
+                return { edits: getInlineHereEdits(context, declaration, selectedUsage!) };
+            default:
+                return Debug.fail("invalid action");
+        }
+    }
+
+    function getInlineAllEdits(
+        context: RefactorContext,
+        declaration: InlineableFunction,
+        usages: ReadonlyArray<CallExpression>
+    ): FileTextChanges[] {
+        const { file } = context;
+        return textChanges.ChangeTracker.with(context, t => {
+            forEach(usages, oldNode => {
+                inlineAt(context, t, oldNode, declaration);
+            });
+            t.delete(file, declaration);
+        });
+    }
+
+    function getInlineHereEdits(
+        context: RefactorContext,
+        declaration: InlineableFunction,
+        selectedUsage: CallExpression
+    ): FileTextChanges[] {
+        return textChanges.ChangeTracker.with(context, t => {
+            inlineAt(context, t, selectedUsage, declaration);
+        });
+    }
+
+    function inlineAt(
+        context: RefactorContext,
+        t: textChanges.ChangeTracker,
+        targetNode: CallExpression,
+        declaration: InlineableFunction
+    ) {
+        const { file, program } = context;
+        const { parameters, body } = declaration;
+        const checker = program.getTypeChecker();
+        const statement = findAncestor(targetNode, isStatement)!;
+        const renameMap = getConflictingNames(checker, declaration, targetNode);
+
+        const isVoid = returnTypeIsVoidLike(checker, declaration);
+        const returns = findDescendants(
+            body!,
+            n => isReturnStatement(n) && getEnclosingBlockScopeContainer(n) === declaration
+        );
+        const nofReturns = returns.length;
+
+        const statements = [] as Statement[];
+        statements.push(...getVariableDeclarationsFromParameters(parameters, targetNode, transformVisitor));
+
+        const variableStatement = getVariableDeclarationFromReturn(file, declaration, nofReturns, isVoid);
+        if (variableStatement) statements.push(variableStatement);
+
+        const transformedBody = visitEachChild(body!, transformVisitor, nullTransformationContext);
+        statements.push(...filter(transformedBody.statements, s => !isReturnStatement(s)));
+
+        forEach(statements, st => { t.insertNodeBefore(file, statement, st); });
+
+        const returnexpression = getReturnExpression();
+        if (returnexpression) {
+            t.replaceNode(file, targetNode, returnexpression);
+        }
+        else {
+            t.deleteRange(file, { pos: statement.getStart(), end: statement.getEnd() });
+        }
+
+        function getReturnExpression() {

```

Comment:
is it possible to extract getReturnExpression, transformVisitor, getSafeName to outer scope?

Reply:
visitor can't be extracted because of restrictions.